src/excmds.ts

```diff
@@ -236,10 +236,24 @@ export async function tabclose(ids?: number[] | number) {
     }
 }
 
+// restore most recently closed tab in this window unless the most recently closed item was a window
 //#background
 export async function undo(){
-    let id = (await browser.sessions.getRecentlyClosed({maxResults: 1}))[0].id
-    browser.sessions.restore(id)
+    const current_win_id : number = (await browser.windows.getCurrent()).id
+    const win = (await browser.sessions.getRecentlyClosed({}))
```

Comment:
The filter is optional, so you don't need the empty object.

Typescript complains, but that's only because our definition of the browser.sessions interface is wrong, you can fix it in tridactyl.d.ts (and I'll eventually merge it into the web-ext-types npm module).

Reply:
More importantly, the object you get back isn't a window, it's an array of session objects, so you should use a different name for this variable

Reply:
My rewrite:

```typescript
/** restore most recently closed tab in this window unless the most recently closed item was a window */
//#background
export async function undo(){
    const current_win_id : number = (await browser.windows.getCurrent()).id
    const sessions = await browser.sessions.getRecentlyClosed()

    // The first session object that's a window or a tab from this window. Or undefined if sessions is empty.
    let closed = sessions.find((s) => {
        return (s.window || s.tab && (s.tab.windowId == current_win_id))
    })
    if (closed) {
        if (closed.tab) {
            browser.sessions.restore(closed.tab.sessionId)
        }
        else if (closed.window) {
            browser.sessions.restore(closed.window.sessionId)
        }
    }
}
```

Note the change in style of the function comment (two leading stars tell typedoc it's a doccomment), the lack of parentheses around the second await (don't need them unless you want to address a member of the awaited object), and the changes to variable names